pkg/network/master/subnets.go

```diff
@@ -238,54 +235,55 @@ func (master *OsdnMaster) handleDeleteNode(obj interface{}) {
 
 // Watch for all hostsubnet events and if one is found with the right annotation, use the SubnetAllocator to dole a real subnet
 func (master *OsdnMaster) watchSubnets() {
-	common.RunEventQueue(master.networkClient.Network().RESTClient(), common.HostSubnets, func(delta cache.Delta) error {
-		hs := delta.Object.(*networkapi.HostSubnet)
-		name := hs.ObjectMeta.Name
-		hostIP := hs.HostIP
-		subnet := hs.Subnet
+	common.RegisterSharedInformer(master.informers, master.handleAddOrUpdateSubnet, master.handleDeleteSubnet, common.HostSubnets)
+}
 
-		log.V(5).Infof("Watch %s event for HostSubnet %q", delta.Type, hs.ObjectMeta.Name)
-		switch delta.Type {
-		case cache.Sync, cache.Added, cache.Updated:
-			if _, ok := hs.Annotations[networkapi.AssignHostSubnetAnnotation]; ok {
-				// Delete the annotated hostsubnet and create a new one with an assigned subnet
-				// We do not update (instead of delete+create) because the watchSubnets on the nodes
-				// will skip the event if it finds that the hostsubnet has the same host
-				// And we cannot fix the watchSubnets code for node because it will break migration if
-				// nodes are upgraded after the master
-				err := master.networkClient.Network().HostSubnets().Delete(name, &metav1.DeleteOptions{})
-				if err != nil {
-					log.Errorf("Error in deleting annotated subnet from master, name: %s, ip %s: %v", name, hostIP, err)
-					return nil
-				}
-				var hsAnnotations map[string]string
-				if vnid, ok := hs.Annotations[networkapi.FixedVNIDHostAnnotation]; ok {
-					vnidInt, err := strconv.Atoi(vnid)
-					if err == nil && vnidInt >= 0 && uint32(vnidInt) <= network.MaxVNID {
-						hsAnnotations = make(map[string]string)
-						hsAnnotations[networkapi.FixedVNIDHostAnnotation] = strconv.Itoa(vnidInt)
-					} else {
-						log.Errorf("Vnid %s is an invalid value for annotation %s. Annotation will be ignored.", vnid, networkapi.FixedVNIDHostAnnotation)
-					}
-				}
-				_, err = master.addNode(name, hostIP, hsAnnotations, nil)
-				if err != nil {
-					log.Errorf("Error creating subnet for node %s, ip %s: %v", name, hostIP, err)
-					return nil
-				}
-			}
-		case cache.Deleted:
-			if _, ok := hs.Annotations[networkapi.AssignHostSubnetAnnotation]; !ok {
-				// release the subnet
-				_, ipnet, err := net.ParseCIDR(subnet)
-				if err != nil {
-					return fmt.Errorf("error parsing subnet %q for node %q for deletion: %v", subnet, name, err)
-				}
-				for _, possibleSubnetAllocator := range master.subnetAllocatorList {
-					possibleSubnetAllocator.ReleaseNetwork(ipnet)
-				}
+func (master *OsdnMaster) handleAddOrUpdateSubnet(obj, _ interface{}, eventType watch.EventType) {
+	hs := obj.(*networkapi.HostSubnet)
+	log.V(5).Infof("Watch %s event for HostSubnet %q", eventType, hs.Name)
+
+	if _, ok := hs.Annotations[networkapi.AssignHostSubnetAnnotation]; ok {
```

Comment:
flip this; just return from the function early if the annotation is not present. (likewise in handleDeleteSubnet)